### PR TITLE
Problem: signature verification result not cache between incarnations of same tx

### PR DESCRIPTION
**Pull request by yihuang**
Closes: #564

Solution:
- introduce incarnation cache that's shared between incarnations of the same tx

# Description

Closes: #XXXX

---

## Author Checklist

*All items are required. Please add a note to the item if the item is not applicable and
please add links to any relevant follow up issues.*

I have...

* [ ] included the correct [type prefix](https://github.com/commitizen/conventional-commit-types/blob/v3.0.0/index.json) in the PR title
* [ ] confirmed `!` in the type prefix if API or client breaking change
* [ ] targeted the correct branch (see [PR Targeting](https://github.com/cosmos/cosmos-sdk/blob/main/CONTRIBUTING.md#pr-targeting))
* [ ] provided a link to the relevant issue or specification
* [ ] reviewed "Files changed" and left comments if necessary
* [ ] included the necessary unit and integration [tests](https://github.com/cosmos/cosmos-sdk/blob/main/CONTRIBUTING.md#testing)
* [ ] added a changelog entry to `CHANGELOG.md`
* [ ] updated the relevant documentation or specification, including comments for [documenting Go code](https://blog.golang.org/godoc)
* [ ] confirmed all CI checks have passed

## Reviewers Checklist

*All items are required. Please add a note if the item is not applicable and please add
your handle next to the items reviewed if you only reviewed selected items.*

I have...

* [ ] confirmed the correct [type prefix](https://github.com/commitizen/conventional-commit-types/blob/v3.0.0/index.json) in the PR title
* [ ] confirmed all author checklist items have been addressed
* [ ] reviewed state machine logic, API design and naming, documentation is accurate, tests and test coverage

### CODE REVIEW

File: CHANGELOG.md [hunk 1]
```diff
@@ -48,6 +48,7 @@ Ref: https://keepachangelog.com/en/1.0.0/
 * [#252](https://github.com/crypto-org-chain/cosmos-sdk/pull/252) Add `BlockGasWanted` to `Context` to support feemarket module.
 * [#269](https://github.com/crypto-org-chain/cosmos-sdk/pull/269) Add `StreamingManager` to baseapp to extend the abci listeners.
 * (crypto/keyring) [#20212](https://github.com/cosmos/cosmos-sdk/pull/20212) Expose the db keyring used in the keystore.
+* (baseapp) [#]() Support incarnation cache when executed in block-stm.
```
Comment (yihuang) [inline CHANGELOG.md]:
```suggestion
* (baseapp) [#565](https://github.com/crypto-org-chain/cosmos-sdk/pull/565) Support incarnation cache when executed in block-stm.
```

File: types/context.go [hunk 2]
```diff
@@ -75,6 +75,9 @@ type Context struct {
 	blockGasUsed uint64
 	// sum the gas wanted by all the transactions in the current block, only accessible by end blocker
 	blockGasWanted uint64
+
+	// incarnationCache is shared between multiple incranations of the same transaction.
```
Comment (yihuang) [inline types/context.go]:
```suggestion
	// incarnationCache is shared between multiple incarnations of the same transaction,
	// it must only cache stateless computation results, like the result of tx signature verification.
```

File: types/context.go [hunk 3]
```diff
@@ -75,6 +75,10 @@ type Context struct {
 	blockGasUsed uint64
 	// sum the gas wanted by all the transactions in the current block, only accessible by end blocker
 	blockGasWanted uint64
+
+	// incarnationCache is shared between multiple incarnations of the same transaction,
+	// it must only cache stateless computation results, like the result of tx signature verification.
```
Comment (yihuang) [inline types/context.go]:
```suggestion
	// it must only cache stateless computation results that only depends on tx body and block level information that don't change during block execution, like the result of tx signature verification.
```

File: x/auth/ante/sigverify.go [hunk 4]
```diff
@@ -330,12 +332,24 @@ func (svd SigVerificationDecorator) AnteHandle(ctx sdk.Context, tx sdk.Tx, simul
...
 	}
+	return nil
+}
 
+func (svd SigVerificationDecorator) AnteHandle(ctx sdk.Context, tx sdk.Tx, simulate bool, next sdk.AnteHandler) (newCtx sdk.Context, err error) {
+	if v, ok := ctx.GetIncarnationCache(SigVerificationResultCacheKey); ok {
+		err = v.(error)
+	} else {
+		err = svd.anteHandle(ctx, tx, simulate)
+		ctx.SetIncarnationCache(SigVerificationResultCacheKey, err)
```
Comment (mmsqe) [inline x/auth/ante/sigverify.go:348]: not sure why set err here

File: x/auth/ante/sigverify.go [hunk 5]
```diff
@@ -330,12 +332,24 @@ func (svd SigVerificationDecorator) AnteHandle(ctx sdk.Context, tx sdk.Tx, simul
...
 
 			}
 		}
 	}
+	return nil
+}
 
+func (svd SigVerificationDecorator) AnteHandle(ctx sdk.Context, tx sdk.Tx, simulate bool, next sdk.AnteHandler) (newCtx sdk.Context, err error) {
+	if v, ok := ctx.GetIncarnationCache(SigVerificationResultCacheKey); ok {
+		err = v.(error)
```
Comment (mmsqe) [inline x/auth/ante/sigverify.go]: can add v != nil as well

**Review by mmsqe — APPROVED** [review]

**Pull request merged**